### PR TITLE
Initial changes for supporting both 32 and 64 bit platforms

### PR DESCRIPTION
**Pull request by shadaj**
This PR is the first step in bringing support for multiple platforms (32bit/64bit, Intel/ARM) to Scala Native. It introduces a new object `CrossPlatform`, which contains a method `cross3264` for handling different logic on each platform and a type `Cross3264[A, B]` for handling different types on each platform.

### CODE REVIEW

**Comment by ekrich** [comment]
@shadaj What you have right now looks very understandable. What is the plan to make `CrossPlatform` select 32 vs 64? The CI failed due to `scalafmt`.

**Comment by shadaj** [comment]
@ekrich I'm planning to have separate versions of scalalib, javalib, etc for each platform (and CrossPlatform would be defined according to the target platform), and use the cross project plugin to handle compiling libraries for multiple platforms.

**Comment by shadaj** [comment]
Only 61 failing unit tests left after statically linking libunwind!

**Comment by shadaj** [comment]
@ekrich because Scala Native uses it's own struct type for `stat`, which is hard coded to be a `long long`, I decided to define a separate type there and change the `unistd` type to `CLong` which matches the original definition.

**Comment by shadaj** [comment]
Only five unit tests left! I'm running into some trouble with stack traces, which aren't displaying any Scala Native function names (only "__libc_start_main" is identified). Any ideas on why that may be happening?

**Comment by shadaj** [comment]
Just one failing test now! @densh calling `getsockopt` for `SO_TIMEOUT` is resulting in an `EINVAL` error. Do you know what may be happening?

**Comment by ekrich** [comment]
Your last change to `.travis.yml` has the same script each time for testing except for the three `env` vars. Is there some way to reuse like add a script to the scripts directory?  Can you use the `before-script` still for the CLA instead of putting it in the compile section - or maybe a check-cla script? Are you aware of this initiative? https://github.com/scala-native/scala-native/issues/1060 We have Travis for now, I thought you should be aware anyway.

**Comment by ekrich** [comment]
@shadaj I was looking into the `EINVAL` problem. Is this happening in only release mode or in debug mode too?

**Comment by ekrich** [comment]
@shadaj I think your recent changes look pretty good. In the compile phase I see some output that looks a little strange. 
`/home/travis/build/scala-native/scala-native/nativelib/_arch-x86_64/target/scala-2.11/nativelib_native0.3.8-SNAPSHOT_x86_64_2.11-0.3.8-SNAPSHOT.pom`

**Comment by shadaj** [comment]
@densh one question for how we want to lay out cross-architecture support. Right now, projects are cross built by the full architecture: x86_64, i386, ARM, ARM64, etc. I'm worried that as the number of supported architectures grows this may become hard for libraries to support, and so was wondering if maybe we should instead only allow different types for 32bit vs 64bit (so only two cross projects), and only allow *logic* switching based on other information about the platform (Intel vs ARM for example) so that it can be done at link time and creation of more cross projects is not required.

**Comment by LeeTibbert** [comment]
@shadaj 
The error I have been getting in ServerSocketSuite was actually in the `close` test just _before_ the `soTimeout` test. I find error reporting in the test framework  to be rather imprecise. Could we be
experiencing the same bug?

I am running in `debug` mode on X86 32 bit, no libunwind.  The error only showed up in the test suite. I could not replicate it in a simple sandbox/Test.scala.  

I think the suite is provoking & validly reporting an underlying code defect.  

If you think this might be the problem you are seeing, I can send you a snippet for
javalib/src/main/scala/java/net/ServerSocket.scala. Or I can post it here or someplace you suggest.

I can now run all tests not involving a stack trace on X86 (thanks to your original 0.3.6 changes).

Once I have had some rest and collected my thoughts, I hope to create one or more PRs to address a 
number of issues I discovered in this excursion.

My next step is to see if this change fixes `nativeMode := "release"`.

Fun in the sun

**Comment by LeeTibbert** [comment]
@shadaj , @ekrich 

Anent the failing ServerSocketSuite, I believe that I have discovered the underlying fault & have a fix working in my devo sandbox.  I described the situation in comments for #1128.

As described above, I believe that close() has a problem, but I believe that the ServerSocketSuite issue that we all have been experiencing exists in PlainSocketImpl accept() when a timeout is given. 

I have not been seeing the EINVAL described above.

Hope this helps.

Lee

**Comment by LeeTibbert** [comment]
@shadaj   anent failing soTimeout test.  FYI, for the longer run.  I fixed a series of bugs with
select that sound like the failure test you commented out. The fix is in PR #1248 and the earlier
PRs it requires./builds_upon.  Because of the imprecise reporting of test failures, for me problem occurred not in seting the soTimeout, but in any test with a call to select() with a timeout specified.
This was a problem was platform independent.   

For #1248, I punted and used poll() instead.  In a later PR #1253, I fixed the root cause: obvious but wrong  select() declarations which lead someone astray.

On my I386 hardware, both the Socket & ServerSocket tests succeed, both for nativeMode
"release" and "debug". Because it such a special system the light bulb also works ;-)

I may have some 32 bit pointer substraction (sic) changes in my private build which may 
affect things.  I have not had time yet to see if those changes are in your stream. If not
how is a good way to get them to you?

Thank you for your efforts on 32 bit support.

Cheers,

Lee

**Comment by shadaj** [comment]
@densh tests are all green now! Could we explore merging this in with an experimental warning for 32bit support (only one unit test currently fails)?

File: project/CrossArchitecture.scala [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import sbtcrossproject._
+import scala.scalanative.build
+import build.{Bits, ThirtyTwo, SixtyFour, TargetArchitecture}
+import scala.scalanative.sbtplugin.ScalaNativeCrossVersion
+import scala.scalanative.sbtplugin.ScalaNativePlugin
+import scala.scalanative.sbtplugin.ScalaNativePlugin.autoImport._
+
+case class CrossBitsPlatform(bits: Bits, lib: Boolean = false)
+    extends Platform {
+  def identifier: String = "_bits-" + bits.toString
```
Comment (densh) [inline project/CrossArchitecture.scala]: People often use `x32` and `x64` to distinguish two types of architectures. It's more concise and just as clear.

File: project/build.sbt [hunk 2]
```diff
@@ -21,6 +21,7 @@ addSbtPlugin("com.eed3si9n"       % "sbt-dirty-money"   % "0.1.0")
 addSbtPlugin("org.foundweekends"  % "sbt-bintray"       % "0.5.2")
 addSbtPlugin("com.jsuereth"       % "sbt-pgp"           % "1.0.0")
 addSbtPlugin("com.typesafe"       % "sbt-mima-plugin"   % "0.1.14")
+addSbtPlugin("org.portable-scala" % "sbt-crossproject"  % "0.4.0")
```
Comment (densh) [inline project/build.sbt]: `0.5.0`.

File: project/CrossArchitecture.scala [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import sbt._
+import Keys.{name, unmanagedResourceDirectories, baseDirectory}
+import sbtcrossproject._
+import scala.scalanative.build
+import build.{Bits, ThirtyTwo, SixtyFour, TargetArchitecture}
+import scala.scalanative.sbtplugin.ScalaNativeCrossVersion
+import scala.scalanative.sbtplugin.ScalaNativePlugin
+import scala.scalanative.sbtplugin.ScalaNativePlugin.autoImport._
+
+case class CrossBitsPlatform(bits: Bits, lib: Boolean = false)
```
Comment (densh) [inline project/CrossArchitecture.scala]: Given that we can only ever have two instances, lets make this sealed with two objects inheriting from it.

File: project/CrossArchitecture.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+          case SixtyFour => TargetArchitecture.x86_64
+        })
+      )
+  }
+
+  @deprecated("Will be removed", "0.3.0")
+  val crossBinary: CrossVersion = ScalaNativeCrossVersion.binary(bits)
+
+  @deprecated("Will be removed", "0.3.0")
+  val crossFull: CrossVersion = ScalaNativeCrossVersion.full(bits)
```
Comment (densh) [inline project/CrossArchitecture.scala]: These two deprecated members are gone in cross project `0.5.0`.

File: project/CrossArchitecture.scala [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+  override def equals(other: Any) = other match {
+    case CrossBitsPlatform(arch, _) => bits == arch
+    case _                          => false
+  }
```
Comment (densh) [inline project/CrossArchitecture.scala]: This method would be unnecessary with two canonical instances.

**Comment by densh** [comment]
Also needs a rebase due to conflicting changes on master.

File: nativelib/_bits-32/src/main/scala/scala/scalanative/runtime/CrossPlatform.scala [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package scala.scalanative
+package runtime
+
+object CrossPlatform {
+  @inline final def cross3264[A, B](thirtyTwo: => A,
```
[the remark this answers is not in the document]
Comment (shadaj) [reply nativelib/_bits-32/src/main/scala/scala/scalanative/runtime/CrossPlatform.scala]: The reason I went with this method was to handle different types across 32/64 bit platforms, which is needed in places like `nativelib`.

**Comment by shadaj** [comment]
Superseded by https://github.com/scala-native/scala-native/pull/1363

**Pull request closed without merging**